### PR TITLE
🐛 fix mysql-cdc OOM: use capped queue

### PR DESCRIPTION
**Pull request by subodh1810**
Issue : https://github.com/airbytehq/airbyte/issues/3969
Read https://github.com/airbytehq/airbyte/issues/3969#issuecomment-864007649 for detailed understanding of the issue

By using a queue with capped capacity we avoid inserting new elements in the queue until there is space for new elements. This avoids heap getting bigger

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Secrets are annotated with `airbyte_secret` in output spec
- [ ] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [ ] Documentation updated 
    - [ ] README
    - [ ] CHANGELOG.md
    - [ ] Reference docs in the `docs/integrations/` directory.
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

<details><summary> <strong> Connector Generator checklist </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Review by tuliren — APPROVED** [review]

**Comment by cgardens** [comment]
`DebeziumEngine` follows a consumer / producer model (like Kafka). My understanding of this change is that you are slowing down the consumer (which takes a record and puts in the queue), but not the producer (thing actually get logs from db). Does slowing down the consumer actually slow down the producer in this case?

**Comment by davinchia** [comment]
Huh interesting. Should we make a similar change on the Postgres source?

**Review by davinchia — APPROVED** [review]
I want to wait for @cgardens 's comment to be resolved - though it sounds like slowing down the producer is noop since the producer is bottlenecked on consumption.

We should have a follow up ticket to double check this for Postgres CDC.

**Comment by subodh1810** [comment]
@cgardens debezium also maintains an internal queue of records but it is also bounded by capacity and has a fixed size. So if we slow down the consumer, producer would also be slowed down

**Comment by subodh1810** [comment]
created issue for postgres https://github.com/airbytehq/airbyte/issues/4253

File: airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/DebeziumRecordPublisher.java [hunk 1]
```diff
@@ -86,7 +86,10 @@ public void start(Queue<ChangeEvent<String, String>> queue) {
           // more on the tombstone:
           // https://debezium.io/documentation/reference/configuration/event-flattening.html
           if (e.value() != null) {
-            queue.add(e);
+            boolean inserted = false;
+            while (!inserted) {
+              inserted = queue.offer(e);
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/DebeziumRecordPublisher.java:91]: should we do a small `sleep` here to ensure we don't busy-loop inside this thread?

File: airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/MySqlSource.java [hunk 2]
```diff
@@ -233,7 +233,7 @@ private static boolean shouldUseCDC(ConfiguredAirbyteCatalog catalog) {
       final AirbyteFileOffsetBackingStore offsetManager = initializeState(stateManager);
       AirbyteSchemaHistoryStorage schemaHistoryManager = initializeDBHistory(stateManager);
       FilteredFileDatabaseHistory.setDatabaseName(config.get("database").asText());
-      final LinkedBlockingQueue<ChangeEvent<String, String>> queue = new LinkedBlockingQueue<>();
+      final LinkedBlockingQueue<ChangeEvent<String, String>> queue = new LinkedBlockingQueue<>(10000);
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/MySqlSource.java:236]: any context behind this magic number? (can we add a comment(

**Review by sherifnada — APPROVED** [review]

File: airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/DebeziumRecordPublisher.java (hunk 1 above)
Comment (subodh1810) [reply airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/DebeziumRecordPublisher.java:91]: @sherifnada What do you think would be a right sleep value?

Comment (sherifnada) [reply airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/DebeziumRecordPublisher.java:91]: ideally expo backoff for 1/2/4/8/16 ms maybe capping at half a second. But barring that a magic constant (erring on the smaller side) should do. For example 10ms.

**Comment by cgardens** [comment]
> @cgardens debezium also maintains an internal queue of records but it is also bounded by capacity and has a fixed size. So if we slow down the consumer, producer would also be slowed down

awesome! thanks for digging into this. can you add a comment explaining this?

File: airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/DebeziumRecordPublisher.java [hunk 3]
```diff
@@ -86,7 +86,15 @@ public void start(Queue<ChangeEvent<String, String>> queue) {
...
+              try {
+                Thread.sleep(10);
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/DebeziumRecordPublisher.java:93]: btw this should only trigger if `inserted == false`

**Review by cgardens — APPROVED** [review]

File: airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/DebeziumRecordPublisher.java (hunk 3 above)
Comment (subodh1810) [reply airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/DebeziumRecordPublisher.java:93]: Ah good catch! sorry about that!

**Comment by subodh1810** [comment]
Merging this PR cause the integration tests pass and the other tests pass locally, we are having issues with Acceptance test

**Pull request merged**